### ISSUE TITLE
integration with caqtdm

### ISSUE THREAD

**Issue by dschick** [taurus-org/taurus#880]
Hej,

I am using Sardana+Tango together with EPICS. For the latter I already have some caQtDM panels for Monitoring my hardware/environment.
I also played a bit with taurus which of course also allows to access the EPICS PVs.

However, I was wondering if it would be possible to have a combined .ui file for both caQtDM and taurus and run it access both protocols?

Right now I do have both plugins installed in the designer so I can place both types of elements into the ui, but then I need to run the .ui file either with caQtDM or taurus.

Best

Daniel

**Comment by cpascual** [comment]
Hi, 

## TL;DR: 
I don't know

## Long (but probably useless) reply

First, please apologize if I am missing something due to my total lack of knowledge of caQtDM... 

From what I understood from the its [project page](https://github.com/caqtdm/caqtdm), caQtDM is a C++ Qt application which interprets the .ui files to generate a C++ application.

Note that taurus widgets are PyQt widgets (not Qt/C++) ones. Other than that, they are pretty standard widgets. You do not need to use a "taurus" application to use a taurus widget. You can use a taurus widget in a regular PyQt application.

If caQtDM (or some other related project) can somehow be used to generate widgets which can be imported and used from a PyQt application, then that would be one way to use both simultaneously.

Now, what about the other way around (using a taurus widget within a caQtDM application)?

The first thing that is not clear to me is **how** does caQtDM use the .ui file: does it load the .ui file using [the standard Qt/C++ mechanisms](https://doc.qt.io/qt-5/designer-using-a-ui-file.html)? or does it use its own custom parsing?

- If it uses the regular Qt mechanisms for loading the .ui file then it is "just" a matter of being able to embed a PyQt widget in a Qt application. I've always thought that this was "easy", but I've never done it, and a quick search on google didn't give me any simple recipe (so maybe it is not easy after all).

- If it does its own parsing, then I guess it is up to caQtDM to add the required support for loading PyQt widgets 

In any case, my point is that if there is a way of using a PyQt widget in caQtDM, then using a taurus widget should work just the same.

PS: it is very nice to hear from uses of Taurus with non-tango systems

**Comment by dschick** [comment]
hej @cpascual ,

thanks a lot for the detailed explaination.
I am actually also not an expert with caQtDM but I might have some time to get into it.
I'll keep you updated.

Daniel

**Comment by cpascual** [comment]
No worries,

I'll close this for now. Do not hesitate to reopen it if you have any further doubt or info.
